File: aquicksort.py
#quicksort implementation in python with array
def quicksort_array(arr):

#1. choosing a pivot element
#2. partitioning the array into elements less than equal to and greater than the pivot
#3. recursively sorting the partitions
    
#time complexity: O(n log n) average O(n²) worst case
#space complexity: O(log n) average for recursion stack

    # base case: if array has 0 or 1 element it's already sorted
    if len(arr) <= 1:
        return arr

    # choose pivot (middle element)
    pivot = arr[len(arr) // 2]
    
    # partition: divide array into three parts
    left = [x for x in arr if x < pivot]   # elements smaller than pivot
    mid  = [x for x in arr if x == pivot]  # elements equal to pivot
    right = [x for x in arr if x > pivot]  # elements greater than pivot

    # recursively sort left and right partitions, then combine
    return quicksort_array(left) + mid + quicksort_array(right)

File: aquicksort.java
//quicksort implementation in java with arrays
public class aquicksort {

    public static void quicksort(int[] arr, int low, int high) {
        if (low < high) {
            // partition the array and get the pivot index
            int p = partition(arr, low, high);
            // recursively sort elements before and after partition
            quicksort(arr, low, p - 1);
            quicksort(arr, p + 1, high);
        }
    }
    private static int partition(int[] arr, int low, int high) {
        // choose the last element as pivot
        int pivot = arr[high];
        int i = low - 1;  // index of smaller element (indicates right position of pivot)

        // traverse through the array and compare each element with pivot
        for (int j = low; j < high; j++) {
            // if current element is smaller than or equal to pivot
            if (arr[j] < pivot) {
                i++;
                swap(arr, i, j);
            }
        }

        // place pivot in its correct position
        swap(arr, i + 1, high);
        return i + 1;
    }
    private static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}
